library/std/src/sys/path/cygwin.rs

```diff
@@ -0,0 +1,67 @@
+use crate::ffi::{OsStr, OsString};
+use crate::os::unix::ffi::OsStringExt;
+use crate::path::{Path, PathBuf, Prefix};
+use crate::sys::common::small_c_string::run_path_with_cstr;
+use crate::sys::cvt;
+use crate::{io, ptr};
+
+#[inline]
+pub fn is_sep_byte(b: u8) -> bool {
+    b == b'/' || b == b'\\'
+}
+
+#[inline]
+pub fn is_verbatim_sep(b: u8) -> bool {
+    b == b'/' || b == b'\\'
+}
+
+#[inline]
+pub fn parse_prefix(_: &OsStr) -> Option<Prefix<'_>> {
+    None
+}
+
+pub const MAIN_SEP_STR: &str = "/";
+pub const MAIN_SEP: char = '/';
+
+unsafe extern "C" {
+    // Doc: https://cygwin.com/cygwin-api/func-cygwin-conv-path.html
+    // Src: https://github.com/cygwin/cygwin/blob/718a15ba50e0d01c79800bd658c2477f9a603540/winsup/cygwin/path.cc#L3902
+    // Safety:
+    // * `what` should be `CCP_WIN_A_TO_POSIX` here
+    // * `from` is null-terminated UTF-8 path
+    // * `to` is buffer, the buffer size is `size`.
+    //
+    // Converts a path to an absolute POSIX path, no matter the input is Win32 path or POSIX path.
+    fn cygwin_conv_path(
+        what: libc::c_uint,
+        from: *const libc::c_char,
+        to: *mut u8,
+        size: libc::size_t,
+    ) -> libc::ssize_t;
+}
+
+const CCP_WIN_A_TO_POSIX: libc::c_uint = 2;
+
+/// Make a POSIX path absolute.
+pub(crate) fn absolute(path: &Path) -> io::Result<PathBuf> {
+    run_path_with_cstr(path, &|path| {
+        let size = cvt(unsafe {
+            cygwin_conv_path(CCP_WIN_A_TO_POSIX, path.as_ptr(), ptr::null_mut(), 0)
+        })?;
+        // If success, size should not be 0.
+        debug_assert!(size >= 1);
```

Comment:
It is possible, though not ideal of course, to have a zero-length `Path`. What happens in this case?

Reply:
The docs say `Otherwise, it returns 0 on success, or -1 on error` so maybe this should be `size >= 0`?

Reply:
As a return from `cygwin_conv_path`, size cannot be zero because it includes the terminating `NUL`.  A zero-length path would be size `1`.

Reply:
> The docs say `Otherwise, it returns 0 on success, or -1 on error` so maybe this should be `size >= 0`?

the "Otherwise" is if the incoming `size` parameter is not `0` (ie, the second call below)

Reply:
Oh, right, thank you for the clarification!